2/homework/laughter_prediction/predictors.py

```diff
@@ -74,9 +77,18 @@ def __init__(self, threshold=0.045985743):
 
 
 class RnnPredictor(Predictor):
-    # Your code here
-    def predict(self, X):
-        pass
-
-    def predict_proba(self, X):
-        pass
\ No newline at end of file
+    def __init__(self, model: LibrosaFeaturesRnn):
+        self.model = model
+
+    def predict(self, X, device='cpu', threshold=0.5):
+        P = self.predict_proba(X, device)
+        res = np.zeros_like(P, dtype=np.int)
+        res[P < threshold] = 0
+        res[P >= threshold] = 1
+        return res
+
+    def predict_proba(self, X, device='cpu'):
+        self.model.set_batch_size(len(X))
+        X_tensor = torch.tensor(X).float().to(device)
+        predicted_proba, _ = self.model.forward(X_tensor, device)
+        return torch.exp(predicted_proba[:,:,1]).cpu().detach().numpy()
```

Comment:
I realise this might suggested by my poor choice of predictor interface. Just keep in mind that in pytorch it is ok to return logits logarithms without exponentiating. Also preserves numeric stability